### PR TITLE
PipelineRun with embedded pipelineSpec and taskSpec

### LINKED ISSUE #143: Generate PipelineRun with embedded pipelineSpec and taskSpec
/kind feature

**Description:**
Convert the YAML produced by the `kfp-tekton` compiler:
- from multiple documents (`Condition`s, `Task`s, `Pipeline`, `PipelineRun`) into
- into a single `PipelineRun` with _embedded_  `pipelineSpec` and `taskSpec`s

This will simply the required work for down-stream integration with the KFP API and KFP UI.

**Additional information:**

```YAML
spec:
  pipelineSpec:
    tasks:
    - name: task1
      taskSpec:
        steps:
          ...
```

### PR DESCRIPTION
**Pull request by ckadner**
**Which issue is resolved by this Pull Request:** 
Resolves #143

**Description of your changes:**
Convert the YAML produced by the `kfp-tekton` compiler:
- from multiple documents (`Condition`s, `Task`s, `Pipeline`, `PipelineRun`) into
- into a single `PipelineRun` with _embedded_  `pipelineSpec` and `taskSpec`s

This will simplify the required work for down-stream integration with the KFP API and KFP UI.

**Outstanding Work**:
TBD

**Environment tested:**

* Python Version (use `python --version`): `3.7.5`
* Tekton Version (use `tkn version`): `0.11.0`
* Kubernetes Version (use `kubectl version`): `0.15`
* OS (e.g. from `/etc/os-release`): Mac OS

---

**Test cases that still need to be fixed (6/15/2020):**

- [ ] big_data_passing.yaml
- [x] condition.yaml
   - missing `super-condition`, solved by ```kubectl -n kubeflow apply -f https://raw.githubusercontent.com/kubeflow/kfp-tekton-backend/master/manifests/kustomize/base/tekton/catalog-condition.yaml```
- [ ] input_artifact_raw_value.yaml
   - `unknown field "artifacts"`
- [x] katib.yaml
- [ ] pipelineparams.yaml
   - `[download : main] sh: results.downloaded-resultoutput.path: not found`
- [ ] pipeline_transformers.yaml
   - missing task/pod annotations
- [x] resourceop_basic.yaml
   - `invalid input params: missing values for these params which have no default values: [success-condition failure-condition]`
- [x] volume_op.yaml
   - `invalid input params: missing values for these params which have no default values: [success-condition failure-condition]`
- [x] volume_snapshot_op.yaml

---

**Issues to track the outstanding fixes:**

- [ ] #188: Outstanding fixes from PR166 (Umbrella issue, @ckadner)
- [ ] #189: Need further big_data_passing fixes after PR166 (@fenglixa)
- [ ] #152: Enable PipelineRun and artifact features as default (@Tomcli)

### CODE REVIEW

File: sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py [hunk 1]
```diff
@@ -12,10 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import List, Text, Dict, Any
+# from typing import List, Text, Dict, Any
 
 
-def fix_big_data_passing(workflow: List[Dict[Text, Any]]) -> List[Dict[Text, Any]]:  # Tekton change signature
+def fix_big_data_passing(workflow: dict) -> dict:
```
Comment (fenglixa) [inline sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py:18]: Hi, @ckadner , could you please don't change the type input parma type for fix_big_data_passing? as I already hanled it in another PR with List type, and hanled pipelinerun together. Thanks ckadner very much!

File: sdk/python/kfp_tekton/compiler/compiler.py [hunk 2]
```diff
@@ -482,31 +424,79 @@ def _create_pipeline_workflow(self, args, pipeline, op_transformers=None, pipeli
...
-    pipeline_template = {
+    # TODO: generate the PipelineRun template
+    pipeline_run = {
       'apiVersion': tekton_api_version,
-      'kind': 'Pipeline',
+      'kind': 'PipelineRun',
       'metadata': {
-        'name': pipeline.name or 'Pipeline',
-        'annotations': {
+        'name': sanitize_k8s_name(pipeline.name or 'Pipeline', suffix_space=4) + '-run',
```
Comment (Tomcli) [inline sdk/python/kfp_tekton/compiler/compiler.py:432]: nit: We no longer need the `-run` at the end because there won't by any duplicated pipeline name in the embedded spec.

File: sdk/python/kfp_tekton/compiler/compiler.py [hunk 3]
```diff
@@ -482,31 +424,79 @@ def _create_pipeline_workflow(self, args, pipeline, op_transformers=None, pipeli
...
+        'annotations': {  # TODO: do annotations apply to PipelineRun?
```
Comment (Tomcli) [inline sdk/python/kfp_tekton/compiler/compiler.py:433]: Yes, annotations in pipelinerun do apply to all tasks which is the same behavior.

**Comment by Tomcli** [comment]
I will take a deeper look with the api and see any errors within the integration.

File: sdk/python/kfp_tekton/compiler/compiler.py [hunk 4]
```diff
@@ -482,31 +424,79 @@ def _create_pipeline_workflow(self, args, pipeline, op_transformers=None, pipeli
...
+    if len(pipeline.conf.image_pull_secrets) > 0:
+      service_template = {
+        'apiVersion': 'v1',
+        'kind': 'ServiceAccount',
+        'metadata': {'name': sanitize_k8s_name(pipeline_run['metadata']['name'], suffix_space=3) + '-sa'}
+      }
+    for image_pull_secret in pipeline.conf.image_pull_secrets:
+      service_template['imagePullSecrets'] = [{'name': image_pull_secret.name}]
+
+    # TODO: service_template
```
Comment (Tomcli) [inline sdk/python/kfp_tekton/compiler/compiler.py:491]: Today the image secret pull PR is finally merged in Tekton, you can ignore this for now and I can create a separate PR to fix it with the new implementation.

**Comment by ckadner** [comment]
[tektoncd/pipeline #2767](https://github.com/tektoncd/pipeline/issues/2767): Unsupported features for embedded pipelineSpec, taskSpec?

File: sdk/python/tests/compiler/testdata/volume_op.yaml [hunk 5]
```diff
@@ -13,117 +13,106 @@
...
+              fieldRef:
+                fieldPath: metadata.labels['tekton.dev/pipelineRun']
+          image: $(params.image)
+          name: main
+          resources: {}
+    - name: cop
+      params:
+      - name: create-pvc-name
+        value: $(tasks.create-pvc.results.name)
+      taskSpec:
```
Comment (Tomcli) [inline sdk/python/tests/compiler/testdata/volume_op.yaml:104]: The volumes from task spec are missing.

File: sdk/python/tests/compiler/testdata/volume.yaml [hunk 6]
```diff
@@ -13,68 +13,54 @@
...
-  - name: echo
-    params:
-    - name: download-downloaded
-      value: $(tasks.download.results.downloaded)
-    taskRef:
-      name: echo
+    tasks:
+    - name: download
+      params: []
+      taskSpec:
```
Comment (Tomcli) [inline sdk/python/tests/compiler/testdata/volume.yaml:30]: Same here, missing volumes for taskSpec.

File: sdk/python/tests/compiler/testdata/sidecar.yaml [hunk 7]
```diff
@@ -13,61 +13,46 @@
...
-  - name: echo
-    params:
-    - name: download-downloaded
-      value: $(tasks.download.results.downloaded)
-    taskRef:
-      name: echo
+    tasks:
+    - name: download
+      params: []
+      taskSpec:
```
Comment (Tomcli) [inline sdk/python/tests/compiler/testdata/sidecar.yaml:30]: The sidecar spec is missing in this taskSpec

File: sdk/python/tests/compiler/testdata/pipeline_transformers.yaml [hunk 8]
```diff
@@ -13,54 +13,38 @@
...
-  - name: print
-    params: []
-    taskRef:
-      name: print
-  - name: print-2
+  pipelineSpec:
     params: []
-    taskRef:
-      name: print-2
+    tasks:
```
Comment (Tomcli) [inline sdk/python/tests/compiler/testdata/pipeline_transformers.yaml:28]: Note: this transformer example is one of the missing features for task labels/annotations

**Comment by ckadner** [comment]
Thanks @Tomcli -- I will need to make more changes to address the missing pieces.

**Comment by ckadner** [comment]
@fenglixa -- I tried to update your code in `_data_passing_rewriter.py` but I must have missed a few places or tripped up over one of the many mapping structures. Some of those mappings are no longer needed (i.e. embedded Tasks names) some need to find their targets in other sub-structures now.

Would you mind taking a look? If this requires a lot more effort than a just a few more small-ish changes, it might take a separate PR after this one gets merged (pending other fixes here).

**Comment by fenglixa** [comment]
Hello @ckadner, I am sorry, I don't think such more code diff is needed for `big data passing`, even though the workflow structure has changed in this PR. Why?

1. In my opinion, the main changes of this PR is integrate the **tasks, pipeline, and pipelinerun** together from **list** to a **dict**. So, in _data_passing_rewriter, we just need to know which one is task template, pipeline and pipelinerun. Then should be no much code diff inside of `_data_passing_rewriter.py`.

2. There is not task name in pipelineSpec['tasks']['taskSpec'], so the task name should be used as pipelineSpec['tasks']['name'], same as the tasks defined for pipeline.

3. task['taskRef']['name'] should be replaced to task['name'], etc...

After all, **`big data passing` is more complex, not even kfp-tekton, but also in kfp**, a little changes may lead to a big modification.  

To make this PR go ahead firstly, I think you can ignore big data part, and log another issue to me, I will handle it after your PR commited.

Thanks.
BR
-Feng

**Comment by ckadner** [comment]
> 1. In my opinion, the main changes of this PR is integrate the **tasks, pipeline, and pipelinerun** together from **list** to a **dict**. So, in `_data_passing_rewriter`, we just need to know which one is task template, pipeline and pipelinerun. Then should be no much code diff inside of `_data_passing_rewriter.py`.

Yes, that is what I attempted :-)

> 2. There is not task name in `pipelineSpec['tasks']['taskSpec']`, so the task name should be used as `pipelineSpec['tasks']['name']`, same as the tasks defined for pipeline.

I think I did that, no?

> 3. `task['taskRef']['name']` should be replaced to `task['name']`, etc...

Yup, ... thought I did.

I think where I tripped up, or where I would need much more time is with all of the mapping data structures. Do you want me to keep my changes and you continue, or would you like me to revert everything in `_data_passing_rewriter` for the scope of this PR?

**Comment by fenglixa** [comment]
> Do you want me to keep my changes and you continue, or would you like me to revert everything in `_data_passing_rewriter` for the scope of this PR?

Yes, you could keep your changes, and I will continue on it. Thanks @ckadner

**Comment by ckadner** [comment]
@Tomcli @animeshsingh 

there are some test cases that still need to be fixed. Feng agreed to tackle the `big_data_passing` in a subsequent PR. I am still trying to fix the others but we could opt to do those in (a) separate PR(s) as well:

- [ ] big_data_passing.yaml (Feng agreed to tackle this in a subsequent PR)
- [ ] condition.yaml
   - missing `super-condition`
- [ ] input_artifact_raw_value.yaml
   - `unknown field "artifacts"` ... missing `stepTemplate` with `volumeMounts`
- [ ] pipelineparams.yaml
   - `[download : main] sh: results.downloaded-resultoutput.path: not found`
- [ ] pipeline_transformers.yaml
   - missing task/pod annotations

**Comment by Tomcli** [comment]
@ckadner for condition pipelines, we assumes there will be a super-condition pre-installed in the same namespace. Is it done as part of the KFP kustomize install https://github.com/kubeflow/kfp-tekton-backend/blob/master/manifests/kustomize/base/tekton/catalog-condition.yaml

So for your test, maybe have this catalog-condition.yaml pre-install on your testing namespace.

**Comment by Tomcli** [comment]
For `pipeline_transformers.yaml`, put this as an issue and we can come back to it once Tekton has it implemented.

**Comment by Tomcli** [comment]
The other three pipelines were failed because it sort of depended on Feng's big_data_passing. We can open an issue and take care of it in a follow up PR.

**Comment by Tomcli** [comment]
/lgtm
Right now the artifact is still optional, but we need to make it as default when we move to KFP for metadata, persisted logs, and artifact tracking. I will open an issue to do a follow up with this PR to test artifacts enabled by default.

**Comment by ckadner** [comment]
> @ckadner for condition pipelines, we assume there will be a super-condition pre-installed in the same namespace. It's done as part of the KFP kustomize install https://github.com/kubeflow/kfp-tekton-backend/blob/master/manifests/kustomize/base/tekton/catalog-condition.yaml
> 
> So for your test, maybe have this catalog-condition.yaml pre-install on your testing namespace.

```Bash
kubectl apply -n kubeflow \
    -f https://raw.githubusercontent.com/kubeflow/kfp-tekton-backend/master/manifests/kustomize/base/tekton/catalog-condition.yaml
```

... worked. Thanks

**Comment by ckadner** [comment]
Thank you @Tomcli -- I opened issues to track the outstanding work.

@animeshsingh  -- after @Tomcli 's review now I consider this PR ready to merge, unless you have further comments

**Comment by animeshsingh** [comment]
thanks @ckadner 

@fenglixa on the three mentioned here, would it possible to bring in one single PR, if you agree they depend on big data passing?

- [x] **big_data_passing.yaml** (@fenglixa agreed to tackle this in a subsequent PR)
- [ ] **input_artifact_raw_value.yaml** (might depend on `big_data_passing` fix)
   - `unknown field "artifacts"` ... missing `stepTemplate` with `volumeMounts`
- [ ] **pipelineparams.yaml** (might depend on `big_data_passing` fix)
   - `[download : main] sh: results.downloaded-resultoutput.path: not found`

File: sdk/README.md [hunk 9]
```diff
@@ -17,7 +17,7 @@ to ensure you are set up properly to use the KFP-Tekton compiler.
   - [How to use the KFP-Tekton Compiler](#how-to-use-the-kfp-tekton-compiler)
   - [Build Tekton from Master](#build-tekton-from-master)
   - [Additional Features](#additional-features)
-    - [1. Compile Kubeflow Pipelines as a Tekton PipelineRun](#1-compile-kubeflow-pipelines-as-a-tekton-pipelinerun)
+    - [1. TODO: remove/restructure: Compile Kubeflow Pipelines as a Tekton PipelineRun](#1-compile-kubeflow-pipelines-as-a-tekton-pipelinerun)
```
Comment (animeshsingh) [inline sdk/README.md:20]: TODO: remove/restructure: 
you want to address this?

File: sdk/README.md [hunk 10]
```diff
@@ -107,32 +98,24 @@ Features that require special builds different from the 'Tested Version' will be
 
 ## Additional Features
 
-### 1. Compile Kubeflow Pipelines as a Tekton PipelineRun
+### 1. TODO: remove/reword: Compile Kubeflow Pipelines as a Tekton PipelineRun
```
Comment (animeshsingh) [inline sdk/README.md:101]: TODO: remove/reword: 

probably address now?

File: sdk/README.md [hunk 11]
```diff
@@ -107,32 +98,24 @@ Features that require special builds different from the 'Tested Version' will be
...
 is generated by the `tkn` CLI so that users can interactively change their pipeline parameters during each execution.
 However, `tkn` CLI is lagging several important features when generating `PipelineRun`.
-Therefore, we added support for generating pipelineRun using `dsl-compile-tekton` with all the latest `kfp-tekton` compiler
-features. The comparison between Tekton pipeline and Argo workflow is described in our 
+
+The comparison between Tekton pipeline and Argo workflow is described in our 
 [design docs](https://docs.google.com/document/d/1oXOdiItI4GbEe_qzyBmMAqfLBjfYX1nM94WHY3EPa94/edit#heading=h.f38y0bqkxo87).
 
-Compiling Kubeflow Pipelines into a Tekton `PipelineRun` is currently in the experimental stage.
+The `kfp-tekton` compiler is now generating a `PipelineRun` when using `dsl-compile-tekton`.
```
Comment (animeshsingh) [inline sdk/README.md:110]:
```suggestion
The `kfp-tekton` compiler generates a `PipelineRun` when using `dsl-compile-tekton`.
```

File: sdk/README.md [hunk 12]
```diff
@@ -107,32 +98,24 @@ Features that require special builds different from the 'Tested Version' will be
...
+
 [Here](https://github.com/tektoncd/pipeline/blob/master/docs/pipelineruns.md) is the list of supported features in `PipelineRun`.
 
-As of today, the below `PipelineRun` features are available within `dsl-compile-tekton`:
+As of today, the following `PipelineRun` features are available within `dsl-compile-tekton`:
  - Affinity
  - Node Selector
  - Tolerations
```
Comment (animeshsingh) [inline sdk/README.md:117]: is this still relevant @Tomcli ?

**Comment by ckadner** [comment]
Thanks @animeshsingh -- I took out the section about the need to generate `pipelineRun` to enable Affinity, Node Selector, and Tolerations since after this PR we will always generate a PipelineRun.

**Comment by fenglixa** [comment]
> @fenglixa on the three mentioned here, would it possible to bring in one single PR, if you agree they depend on big data passing?
> 
> * [x]  **big_data_passing.yaml** (@fenglixa agreed to tackle this in a subsequent PR)
> * [ ]  **input_artifact_raw_value.yaml** (might depend on `big_data_passing` fix)
>   
>   * `unknown field "artifacts"` ... missing `stepTemplate` with `volumeMounts`
> * [ ]  **pipelineparams.yaml** (might depend on `big_data_passing` fix)
>   
>   * `[download : main] sh: results.downloaded-resultoutput.path: not found`


@animeshsingh , 1st and 3rd are sure depends on big data passing part. 2nd one about `stepTemplate with volumeMounts missing` I need to further debug to see the detail reasons, maybe related to it as well.

After all, I will address these 3 cases in another single PR together, make sure all of them works (issue #189).

Personally, I think we can let this PR go firstly.

**Comment by fenglixa** [comment]
> Right now the artifact is still optional, but we need to make it as default when we move to KFP for metadata, persisted logs, and artifact tracking. I will open an issue to do a follow up with this PR to test artifacts enabled by default.

As I know,  set `enable_artifacts=True` will lead to watson pipeline failed. I didn't has chance to find out the reason, but we may need take attention on it. @ckadner @Tomcli @animeshsingh

**Comment by Tomcli** [comment]
> > Right now the artifact is still optional, but we need to make it as default when we move to KFP for metadata, persisted logs, and artifact tracking. I will open an issue to do a follow up with this PR to test artifacts enabled by default.
> 
> As I know, set `enable_artifacts=True` will lead to watson pipeline failed. I didn't have chance to find out the reason, but we may need to take attention on it. @ckadner @Tomcli @animeshsingh

Right now, all the pipelines results are sanitizing the identifiers when passing through _big_data_passing. Since we commented out the _big_data_passing in this PR, pipelines with underscore identifier for results (such as the waston ml pipeline) will break on enable_artifacts because it can't map the artifacts with non-sanitize result names. Since there are few edge cases with the artifact identifier that are depended on _big_data_passing, I opened the issue to revisit the artifacts again after _big_data_passing is fixed.

**Comment by animeshsingh** [comment]
@Tomcli @ckadner before we merge, ensuring we have corresponding open issues hyperlinked for the open list here? do we need to have a master issue for "fixes needed after this change" pointing to various open issues here, as well as linking to bigger big data passing related issues?

- [ ] input_artifact_raw_value.yaml
   - `unknown field "artifacts"`
- [x] katib.yaml
- [ ] pipelineparams.yaml
   - `[download : main] sh: results.downloaded-resultoutput.path: not found`
- [ ] pipeline_transformers.yaml
   - missing task/pod annotations
- [x] resourceop_basic.yaml
   - `invalid input params: missing values for these params which have no default values: [success-condition failure-condition]`
- [x] volume_op.yaml
   - `invalid input params: missing values for these params which have no default values: [success-condition failure-condition]`
- [x] volume_snapshot_op.yaml

File: sdk/python/tests/compiler/testdata/resourceop_basic.py [hunk 13]
```diff
@@ -57,7 +57,9 @@ def resourceop_basic():
     op = dsl.ResourceOp(
         name='test-step',
         k8s_resource=json.loads(_CONTAINER_MANIFEST),
-        action='create'
+        action='create',
+        success_condition='status.succeeded > 0',
```
Comment (Tomcli) [inline sdk/python/tests/compiler/testdata/resourceop_basic.py:61]: I just found out you changed this test case, we shouldn't add success_condition and failure_condition because they suppose to be using the default param with `''`.

File: sdk/python/tests/compiler/testdata/resourceop_basic.yaml [hunk 14]
```diff
@@ -13,90 +13,77 @@
...
+        - default: strategic
+          description: Merge strategy when using action patch
+          name: merge-strategy
+          type: string
+        - description: An express to retrieval data from resource.
+          name: output
+          type: string
+        - description: A label selector express to decide if the action on resource
+            is success.
+          name: success-condition
```
Comment (Tomcli) [inline sdk/python/tests/compiler/testdata/resourceop_basic.yaml:56]: Looks like the task default parameter with empty string all get dropped somewhere, this is probably why the resourceop task was failing with the original test.

**Comment by ckadner** [comment]
> @Tomcli @ckadner before we merge, ensuring we have corresponding open issues hyperlinked for the open list here? do we need to have a master issue for "fixes needed after this change" pointing to various open issues here, as well as linking to bigger big data passing related issues?

I had created these issues yesterday, one umbrella-kind issue, one broken out for @fenglixa already.

- #188: Outstanding fixes from PR 166 
- #189: Need further big_data_passing fixes after PR 166

File: sdk/python/tests/compiler/testdata/resourceop_basic.py (hunk 13 above)
Comment (ckadner) [reply sdk/python/tests/compiler/testdata/resourceop_basic.py:61]: @Tomcli -- they failed Tekton validation without it:

```invalid input params: missing values for these params which have no default values: [success-condition failure-condition]```

File: sdk/python/tests/compiler/testdata/resourceop_basic.yaml (hunk 14 above)
Comment (ckadner) [reply sdk/python/tests/compiler/testdata/resourceop_basic.yaml:56]: oh no, the empty string non-default default parameters again :-(

File: sdk/python/tests/compiler/testdata/resourceop_basic.py (hunk 13 above)
Comment (ckadner) [reply sdk/python/tests/compiler/testdata/resourceop_basic.py:61]: @Tomcli -- why are `''` empty string non-default default parameters required here instead of actual `success_condition` and `failure_condition` ?

File: sdk/python/tests/compiler/testdata/resourceop_basic.yaml (hunk 14 above)
Comment (ckadner) [reply sdk/python/tests/compiler/testdata/resourceop_basic.yaml:58]: I pushed another commit, reinstating the empty string default non value default params

File: sdk/python/tests/compiler/testdata/resourceop_basic.py (hunk 13 above)
Comment (Tomcli) [reply sdk/python/tests/compiler/testdata/resourceop_basic.py:61]: We don't have to set the actual success_condition and failure_condition because they should be optional not required. I think the actual issue is at below where the empty string default param are missing.

File: sdk/python/tests/compiler/testdata/resourceop_basic.yaml (hunk 14 above)
Comment (Tomcli) [reply sdk/python/tests/compiler/testdata/resourceop_basic.yaml:58]: It was there in the previous task template, probably one of the clean up functions popped out the empty dictionary value.

File: sdk/python/tests/compiler/testdata/resourceop_basic.py (hunk 13 above)
Comment (Tomcli) [reply sdk/python/tests/compiler/testdata/resourceop_basic.py:61]: Thanks @ckadner, can you revert the resourceop and volumeop test case without the `success_condition` and `failure_condition`. This way we can catch those edge cases if we see errors again.

Comment (ckadner) [reply sdk/python/tests/compiler/testdata/resourceop_basic.py]: I removed the `success_condition` and `failure_condition` again

File: sdk/python/tests/compiler/testdata/resourceop_basic.py [hunk 15]
```diff
@@ -58,8 +58,8 @@ def resourceop_basic():
         name='test-step',
         k8s_resource=json.loads(_CONTAINER_MANIFEST),
         action='create',
```
Comment (Tomcli) [inline sdk/python/tests/compiler/testdata/resourceop_basic.py:60]: remove the extra comma after action?

File: sdk/python/tests/compiler/testdata/volume_op.py [hunk 16]
```diff
@@ -25,8 +25,8 @@ def volumeop_basic(size="10M"):
         resource_name="my-pvc",
         modes=dsl.VOLUME_MODE_RWO,
         size=size,
```
Comment (Tomcli) [inline sdk/python/tests/compiler/testdata/volume_op.py:27]: Same as above

File: sdk/python/tests/compiler/testdata/resourceop_basic.py (hunk 15 above)
Comment (ckadner) [reply sdk/python/tests/compiler/testdata/resourceop_basic.py]: sure

**Comment by Tomcli** [comment]
Thanks @ckadner 
/lgtm

**Comment by ckadner** [comment]
Thanks for your review and fixes @Tomcli 

@animeshsingh -- if you approve to merge this PR, then @fenglixa can start working on the subsequent fixes for `big_data_passing` (#189)

**Comment by animeshsingh** [comment]
@fenglixa just want to get a sense of timing on your side for the related fixes, given big data fix is critical to watson - before i push this in master, since this will break the master for big data fix for a while.

**Comment by fenglixa** [comment]
> @fenglixa just want to get a sense of timing on your side for the related fixes, given big data fix is critical to watson - before i push this in master, since this will break the master for big data fix for a while.

Consider to the watson part,  let me start working on #189 right now, with the trunk of: https://github.com/ckadner/kfp-tekton/tree/issue_143_embedded_pipelineSpec

So let's commit code of this PR with my fixes together later. 

Due to the influence of Chinese Dragon Boat Festival these days, I may need 3-5 natural days to finish it. Let me know if you have any concern on it. Thanks.

**Comment by animeshsingh** [comment]
thanks - to clarify @fenglixa you are recommending to hold this until your PR comes by the end of the week, or merge it now?

**Comment by animeshsingh** [comment]
@ckadner a conflict on this one as well

**Comment by ckadner** [comment]
> @ckadner a conflict on this one as well

Rebased.

**Comment by fenglixa** [comment]
@ckadner @animeshsingh ,
PR submitted for big data part modification.
with my self unit test,  PR #194 + PR #166,  then all the cases should be OK now, including big data. See details of my test result from PR #194

@ckadner,  let's commit these 2 PRs now, to support both pipelinerun and bigdata.
Additional, PR #194 changed following 6 files. To avoid any conflicts during merging, **it's better you can revert these files in this PR,** (it's also OK if you don't want do it, then we may need to resolve conflicts later).

```
	modified:   sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py
	modified:   sdk/python/tests/compiler/testdata/big_data_passing.py
	modified:   sdk/python/tests/compiler/testdata/big_data_passing.yaml
	modified:   sdk/python/tests/compiler/testdata/input_artifact_raw_value.yaml
	modified:   sdk/python/tests/compiler/testdata/katib.yaml
	modified:   sdk/python/tests/compiler/testdata/pipelineparams.yaml
```

**Comment by animeshsingh** [comment]
@ckadner vis a vis the 6 files mentioned above - what do you want to do?

**Comment by ckadner** [comment]
> @ckadner vis a vis the 6 files mentioned above - what do you want to do?

I will keep them since they reflect the conversion from individual `Task`, `Pipeline` to all-inclusice `PipelineRun`.

**Comment by Tomcli** [comment]
/lgtm
Although Feng's PR probably needs to rebase after we merged this PR.

**Comment by ckadner** [comment]
> /lgtm
> Although Feng's PR probably needs to rebase after we merged this PR.

yes. but that should not be a big deal. the testdata gets regenerated and the bigdatafix builds on top of what I had done, and if Feng has no patience he can just override it :-)

**Comment by ckadner** [comment]
oh, and, if I were to take those files out, Travis would complain ;-)

**Comment by animeshsingh** [comment]
/approve

thanks @ckadner for driving this, and @Tomcli @fenglixa for your help. 
cc @Udiknedormin
